torch/csrc/jit/script/compiler.cpp

```diff
@@ -2377,6 +2229,60 @@ struct to_ir {
     return graph->insertConstant(stack[0], nullptr, tree->range());
   }
 
+  // We construct the iterable tree here using the IterableTree SugaredValue,
+  // The tree consists of SimpleValue, RangeValue or IterableValue:
+  // For SimpleValues(List, Dict, etc) or RangeValue. We will make them as tree leaves
+  // since we could get the loop information from len() and get_item().
+  // For IterableValue like zip(), enumerate(), we can model them as a combination of
+  // leaves, and we emit a IterableTree value to record the tree information
+  SugaredValuePtr emitIterableTree(
+      SourceRange& loc,
+      const List<Expr>& inputs,
+      const std::shared_ptr<IterableValue>& iterable) {
+      std::shared_ptr<IterableTree> iterable_tree = nullptr;
+      size_t input_size = inputs.size();
+      if (iterable->symbol_ == prim::enumerate) {
```

Comment:
Could this be done on construction of the sugared value, to avoid having hold a prim::enumerate symbol to direct what to do later?

Reply:
What I observed from the `emitApplyExpr` is that, it will first do `emitSugaredExpr(apply.callee(), 1)` without passing inputs in, and the `builtins` will emit corresponding sugared values, I could not do it there since it's missing the inputs, that's where I need the singleton `IterableValue` sugared value to record the symbol and do the later tree construction. I could instead do a whole special casing on the `IterableTree` when `emitApplyExpr`, but that makes it separate from the `builtin` registration table mechanism that we already using..